src/storage/index/iter.rs

```diff
@@ -0,0 +1,276 @@
+use std::collections::{Bound, VecDeque};
+use std::ops::RangeBounds;
+use std::sync::Arc;
+
+use crate::storage::index::art::{Node, NodeType};
+use crate::storage::index::KeyTrait;
+
+// TODO: need to add more tests for snapshot readers
+/// A structure representing a pointer for iterating over the Trie's key-value pairs.
+pub struct IterationPointer<P: KeyTrait, V: Clone> {
+    pub(crate) id: u64,
+    root: Arc<Node<P, V>>,
+}
+
+impl<P: KeyTrait, V: Clone> IterationPointer<P, V> {
+    /// Creates a new IterationPointer instance.
+    ///
+    /// # Arguments
+    ///
+    /// * `root` - The root node of the Trie.
+    /// * `id` - The ID of the snapshot.
+    ///
+    pub fn new(root: Arc<Node<P, V>>, id: u64) -> IterationPointer<P, V> {
+        IterationPointer { id, root }
+    }
+
+    /// Returns an iterator over the key-value pairs within the Trie.
+    ///
+    /// # Returns
+    ///
+    /// Returns an Iter iterator instance.
+    ///
+    pub fn iter(&self) -> Iter<P, V> {
+        Iter::new(Some(&self.root))
+    }
+
+    pub fn range<'a, R>(
+        &'a self,
+        range: R,
+    ) -> impl Iterator<Item = (Vec<u8>, &'a V, &'a u64, &'a u64)>
+    where
+        R: RangeBounds<P> + 'a,
+    {
+        return Range::new(Some(&self.root), range);
+    }
+}
+
+/// An iterator over the nodes in the Trie.
+struct NodeIter<'a, P: KeyTrait, V: Clone> {
+    node: Box<dyn Iterator<Item = (u8, &'a Arc<Node<P, V>>)> + 'a>,
+}
+
+impl<'a, P: KeyTrait, V: Clone> NodeIter<'a, P, V> {
+    /// Creates a new NodeIter instance.
+    ///
+    /// # Arguments
+    ///
+    /// * `iter` - An iterator over node items.
+    ///
+    fn new<I>(iter: I) -> Self
+    where
+        I: Iterator<Item = (u8, &'a Arc<Node<P, V>>)> + 'a,
+    {
+        Self {
+            node: Box::new(iter),
+        }
+    }
+}
+
+impl<'a, P: KeyTrait, V: Clone> Iterator for NodeIter<'a, P, V> {
+    type Item = (u8, &'a Arc<Node<P, V>>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.node.next()
+    }
+}
+
+/// An iterator over key-value pairs in the Trie.
+pub struct Iter<'a, P: KeyTrait + 'a, V: Clone> {
+    inner: Box<dyn Iterator<Item = (Vec<u8>, &'a V, &'a u64, &'a u64)> + 'a>,
+    _marker: std::marker::PhantomData<P>,
+}
+
+impl<'a, P: KeyTrait + 'a, V: Clone> Iter<'a, P, V> {
+    /// Creates a new Iter instance.
+    ///
+    /// # Arguments
+    ///
+    /// * `node` - An optional reference to the root node of the Trie.
+    ///
+    pub(crate) fn new(node: Option<&'a Arc<Node<P, V>>>) -> Self {
+        if let Some(node) = node {
+            Self {
+                inner: Box::new(IterState::new(node)),
+                _marker: Default::default(),
+            }
+        } else {
+            Self {
+                inner: Box::new(std::iter::empty()),
+                _marker: Default::default(),
+            }
+        }
```

Comment:
I think a `match` statement is a bit easier to read in this case.

```suggestion
        match node {
            Some(node) => Self {
                inner: Box::new(IterState::new(node)),
                _marker: Default::default(),
            },
            None => Self {
                inner: Box::new(std::iter::empty()),
                _marker: Default::default(),
            },
        }
```